Test/paulmck/rcu.hpp

```diff
@@ -3,45 +3,26 @@
 #include <cstddef>
 #include <memory>
 #include <mutex>
+#include <tuple>
 #include <utility>
 
 // Derived-type approach.  All RCU-protected data structures using this
 // approach must derive from std::rcu_obj_base, which in turn derives
 // from std::rcu_head.  No idea what happens in case of multiple inheritance.
 
 namespace std {
-    template<typename T, typename D = default_delete<T>, bool E = is_empty<D>::value>
-    class rcu_obj_base: private rcu_head {
-        D deleter;
+    template<typename T, typename D = default_delete<T>>
+    class rcu_obj_base: private rcu_head, private std::tuple<D> {
```

Comment:
Oh; well, having a member of type `tuple<X,E>` is a bit different from inheriting from `X, tuple<E>`. I'd much prefer that we (which is to say "not me" for the foreseeable future) find a way of triggering the empty-base optimization that definitely *does* work on current GCC, Clang, and MSVC.

I think this `std::tuple` trick works for libstdc++ but not for libc++.
https://wandbox.org/permlink/H0PB0OjM9dH6kVCa